apps/aecore/src/aec_tx_sign.erl

```diff
@@ -11,40 +12,45 @@
 -include("common.hrl").
 -include("txs.hrl").
 
-data(SignedTx) ->
-    SignedTx#signed_tx.data.
+data(#signed_tx{data = Data}) ->
+    Data.
+
+signatures(#signed_tx{signatures = Sigs}) ->
+    Sigs.
 
 verify(failed_tx) ->
-    %% TODO: Verify signed txs
     {error, verification_failed};
-verify(_Tx) ->
-    ok.
+verify(#signed_tx{data = Data, signatures = Sigs}) ->
+    case aec_keys:verify(Sigs, Data) of
+        true -> ok;
+        false ->
+            lager:debug("No matching sigs (~p - ~p)", [Sigs]),
+            {error, signature_check_failed}
+    end.
+
+-define(SIG_TX_TYPE, <<"sig_tx">>).
+-define(SIG_TX_VSN, 1).
+
+version() -> ?SIG_TX_VSN.
+type()    -> ?SIG_TX_TYPE.
 
--spec serialize(#signed_tx{}) -> map().
+-spec serialize(#signed_tx{}) -> list().
 serialize(#signed_tx{data = Tx, signatures = Sigs})  when is_tuple(Tx) ->
-    Mod = tx_dispatcher:handler(Tx),
-    Data = Mod:serialize(Tx),
-    Type = Mod:type(),
-    #{<<"type">> => Type, 
-      <<"data">> => Data,
-      <<"signatures">> => lists:map(fun base64:encode/1, Sigs)}.
-
-
--spec deserialize(map()) -> #signed_tx{}.
-deserialize(#{<<"type">> := Type,
-              <<"signatures">> := Sigs,
-              <<"data">> := Data}) ->
-    Mod = tx_dispatcher:handler_by_type(Type),
-    Tx = Mod:deserialize(Data),
-    #signed_tx{data = Tx,
-               signatures = lists:map(fun base64:decode/1, Sigs)}.
+    TxSer = aec_tx:serialize(Tx),
+    [type(), version(), TxSer, Sigs].
+
+-spec deserialize(list()) -> #signed_tx{}.
+deserialize([?SIG_TX_TYPE, ?SIG_TX_VSN, TxSer, Sigs]) ->
+    Tx = aec_tx:deserialize(TxSer),
+    #signed_tx{data = Tx, signatures = Sigs}.
 
 %% deterministic canonical serialization.
 -spec serialize_to_binary(#signed_tx{}) -> binary().
 serialize_to_binary(#signed_tx{} = SignedTx) ->
-    jsx:encode(serialize(SignedTx)).
+    msgpack:pack(serialize(SignedTx)).
```

Comment:
`aec_tx_sign:serialize_to_binary/1` is used by `aec_blocks:new/3` (via `aec_tx_trees`) for computing [block header `txs_hash` i.e. the "Root hash of the Merkle tree of transactions in the block"](https://github.com/aeternity/epoch/wiki/Block-Header) hence needs to be deterministic and documented (as other nodes in the network need to be able to compute state txs Merkle tree resulting in same root hash).

Could you please:
* Whether this line results in deterministic binary (I understand `jsx:encode(serialize(SignedTx))` was not as JSON can contain an unspecified number of newlines/whitespaces). If out of scope for this PR (as I think it is out of scope) please ensure a ticket is present;
* Open ticket tracking integration of wiki with this root hash computation (i.e. msgpack).

Reply:
As far as I've been able to determine, `msgpack` does produce deterministic binaries. It's a byte-coded protocol, so it's not affected by whitespace or formatting issues.

Reply:
Thanks